lib/std/process.zig

```diff
@@ -967,7 +967,7 @@ pub fn getUserInfo(name: []const u8) !UserInfo {
 /// TODO this reads /etc/passwd. But sometimes the user/id mapping is in something else
 /// like NIS, AD, etc. See `man nss` or look at an strace for `id myuser`.
 pub fn posixGetUserInfo(name: []const u8) !UserInfo {
-    const file = try std.fs.openFileAbsolute("/etc/passwd", .{});
+    const file = try std.fs.cwd().openFile("/etc/passwd", .{});
```

Comment:
I think this is a good example of something that's not an improvement.

The code before will use the `open()` syscall while the code after will use the `openat()` syscall with `AT_FDCWD`. In practice this is generally fine, however, the latter introduces a dependency on the concept of the current working directory. The current working directory, however, is irrelevant to reading this absolute file path.

Another way to put this is that the code before specified the intent more precisely, while the code after specifies intent less precisely.

It is for this reason that I think the "absolute" variants of fs functions should continue to exist.

Reply:
> The code before will use the `open()` syscall

This is not true in this case:

https://github.com/ziglang/zig/blob/f40f81cbfb0a867b32d4406e198550730de268cd/lib/std/fs.zig#L2761-L2764

The only `Absolute` functions that are not purely wrappers around the `Dir` function are:
- readLinkAbsolute, symLinkAbsolute, makeDirAbsolute, deleteDirAbsolute, renameAbsolute
All the rest look just like `openFileAbsolute`.

> Another way to put this is that the code before specified the intent more precisely, while the code after specifies intent less precisely.

I came to the opposite conclusion when writing this PR. From https://github.com/ziglang/zig/issues/16736#issuecomment-1670365838:

> In making the changes in [this] PR, I came across another reason why the `Absolute` versions aren't very useful: the `Absolute` terminology is overloaded and it's hard/impossible to tell why the `Absolute` variant is being used--is it being used just because the programmer happens to know the path is absolute, or is it being used because the programmer knows the path **must** be absolute for the code to work correctly? Removing the `Absolute` variants will remove this ambiguity since the programmer will be forced to make their intentions more explicit if they want to assert/check that a given path is absolute.

Reply:
I'll also add that on some architectures, non `-at` versions of many syscalls don't exist.

Reply:
> All the rest look just like `openFileAbsolute`. For all the changes in this PR, the `strace` would actually be identical before and after.

It would be an improvement if the implementation were changed to match what I said, but it doesn't matter too much if the OS treats it the same.

More importantly, it is an impedance mismatch to call `fs.cwd()` in order to do something with an absolute file path. Looking at the implementation of cwd:

```zig
pub fn cwd() Dir {
    if (builtin.os.tag == .windows) {
        return Dir{ .fd = os.windows.peb().ProcessParameters.CurrentDirectory.Handle };
    } else if (builtin.os.tag == .wasi) {
        return std.options.wasiCwd();
    } else {
        return Dir{ .fd = os.AT.FDCWD };
    }
}
```

We can observe some problems. On Windows, this lowers ultimately to some inline assembly (that is apparently incorrectly marked as having side effects, see #18077 that I just opened) that should not be emitted when accessing an absolute path. On WASI it could cause all sorts of unintended problems because it hooks into a user-overridable function for getting the current working directory.

These things are minor issues, but they are the consequences of specifying the wrong intent.

I don't really understand your second point. The functions dealing with absolute paths assert that the parameter is an absolute path, so I think you have it precisely backwards - using the absolute functions is the way to make intentions more explicit.

Reply:
Thanks, I understand where you're coming from better now. I was mostly focused on the potential downstream benefits of there being one obvious way of doing `fs`-related stuff, but I can understand why avoiding a known-to-be-unnecessary `fs.cwd()` call can be beneficial.

Before I close this, what are your thoughts on this pattern:

https://github.com/ziglang/zig/blob/ac95cfe4496a5504b42bf37b02f34eff390fe956/lib/std/debug.zig#L1077-L1080

Do you see that as something that should be done more, or should that be replaced with `std.fs.cwd().openFile`?

Reply:
I think that pattern should always be replaced with `fs.cwd().openFile`.